ydb/core/tx/datashard/sample_k.cpp

```diff
@@ -0,0 +1,329 @@
+#include "datashard_impl.h"
+#include "range_ops.h"
+#include "scan_common.h"
+#include "upload_stats.h"
+
+#include <ydb/core/base/appdata.h>
+#include <ydb/core/base/counters.h>
+#include <ydb/core/kqp/common/kqp_types.h>
+#include <ydb/core/scheme/scheme_tablecell.h>
+#include <ydb/core/tablet_flat/flat_row_state.h>
+
+#include <ydb/core/tx/tx_proxy/proxy.h>
+#include <ydb/core/tx/tx_proxy/upload_rows.h>
+
+#include <ydb/core/ydb_convert/table_description.h>
+#include <ydb/core/ydb_convert/ydb_convert.h>
+#include <ydb/library/yql/public/issue/yql_issue_message.h>
+
+#include <util/generic/algorithm.h>
+#include <util/string/builder.h>
+
+#define LOG_T(stream) LOG_TRACE_S(*TlsActivationContext, NKikimrServices::TX_DATASHARD, stream)
+#define LOG_E(stream) LOG_ERROR_S(*TlsActivationContext, NKikimrServices::TX_DATASHARD, stream)
+
+namespace NKikimr::NDataShard {
+
+class TSampleKScan final: public TActor<TSampleKScan>, public NTable::IScan {
+protected:
+    const TAutoPtr<TEvDataShard::TEvSampleKResponse> Response;
+    const TActorId ResponseActorId;
+
+    const TTags ScanTags;
+    const TVector<NScheme::TTypeInfo> KeyTypes;
+
+    const TSerializedTableRange TableRange;
+    const TSerializedTableRange RequestedRange;
+    const ui64 K;
+
+    IDriver* Driver = nullptr;
+
+    struct TProbability {
+        ui64 P = 0;
+        ui64 I = 0;
+
+        bool operator==(const TProbability&) const noexcept = default;
+        auto operator<=>(const TProbability&) const noexcept = default;
+    };
+
+    TReallyFastRng32 Rng;
+    ui64 MaxProbability = 0;
+    std::vector<TProbability> MaxRows;
+    std::vector<TString> DataRows;
+
+public:
+    static constexpr NKikimrServices::TActivity::EType ActorActivityType() {
+        return NKikimrServices::TActivity::BUILD_INDEX_SCAN_ACTOR; // TODO(mbkkt) Should be different activity type?
```

Comment:
Yes, it should be a different actitvity.

Reply:
Done